File: Data Structures & Algorithms/palindromic-substrings/submission-0.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        ans = 0
        for idx in range(len(s)):
            c = s[idx]

            left, right = idx - 1, idx + 1
            # odd
            while(left >= 0 and right < len(s) and s[left] == s[right]):
                left -= 1
                right += 1
                ans += 1

            # even
            left, right = idx, idx + 1
            while (left >= 0 and right < len(s) and s[left] == s[right]):
                left -= 1
                right += 1
                ans += 1
        return ans + len(s)

File: Data Structures & Algorithms/longest-palindromic-substring/submission-0.cpp
class Solution {
public:

    string longestLengthPalindromeCenteredAt(string s, int center){

        // Odd length palindrome case
        int left = center - 1;
        int right = center + 1;
        string oddPalindrome = string(1,s[center]);
        while(left >= 0 && right < s.size()){
            if(s[left] != s[right]){
                break;
            }
            oddPalindrome.insert(0, string(1,s[left]));
            oddPalindrome += s[right];
            left--;
            right++;
        }

        // Even length palindrome case
        left = center;
        right = center + 1;
        string evenPalindrome = "";
        while(left >= 0 && right < s.size()){
            if(s[left] != s[right]){
                break;
            }
            evenPalindrome.insert(0, string(1,s[left]));
            evenPalindrome += s[right];
            left--;
            right++;
        }

        // cout << "odd " << oddPalindrome << endl;
        // cout << "even " << evenPalindrome << endl;
        
        return oddPalindrome.size() > evenPalindrome.size() ? oddPalindrome : evenPalindrome;
    }
    string longestPalindrome(string s) {
        string longest = "";
        for(int i = 0; i < s.size(); i++){
            string palin = longestLengthPalindromeCenteredAt(s, i);
            if(palin.size() > longest.size()) {
                longest = palin;
            }
        }
        return longest;
    }
};
